api/src/main/java/jakarta/enterprise/invoke/InvokerBuilder.java

```diff
@@ -0,0 +1,359 @@
+ * during deployment, as described in previous sections. Other types are checked during
+ * invocation, at the very least due to the type checks performed implicitly by the JVM.
+ * The lookups, transformers and possibly the wrapper must arrange the inputs and outputs
+ * so that when the invokable method is eventually invoked, the rules described in
+ * {@link Invoker#invoke(Object, Object[]) Invoker.invoke()} all hold.
+ *
+ * @param <T> type of outcome of this builder; always represents an {@code Invoker},
+ *            but does not necessarily have to be an {@code Invoker} instance directly
+ */
+// TODO more kinds of transformations could be defined, expecially for argument handling
+// TODO it would be possible to specify a sequence of transformations for each input/output, instead of just one
+// TODO we should probably build a prototype for this based on method handles
+public interface InvokerBuilder<T> {
+    /**
+     * Enables lookup of the target instance.
+     *
+     * @return this builder
+     */
+    InvokerBuilder<T> setInstanceLookup();
+
+    /**
+     * Enables lookup of the argument on given {@code position}.
+     *
+     * @param position zero-based argument position for which lookup is enabled
+     * @return this builder
+     * @throws IllegalArgumentException if {@code position} is greather than or equal to
+     * the number of parameters declared by the invokable method
+     */
+    InvokerBuilder<T> setArgumentLookup(int position);
+
+    /**
+     * Configures an input transformer for the target instance.
+     *
+     * @param clazz class that declares the transformer
+     * @param methodName transformer method name
+     * @return this builder
+     * @throws IllegalStateException if this method is called more than once
+     */
+    InvokerBuilder<T> setInstanceTransformer(Class<?> clazz, String methodName);
+
+    /**
+     * Configures an input transformer for the argument on given {@code position}.
+     *
+     * @param position zero-based argument position for which the input transformer is configured
+     * @param clazz class that declares the transformer
+     * @param methodName transformer method name
+     * @return this builder
+     * @throws IllegalArgumentException if {@code position} is greather than or equal to
+     * the number of parameters declared by the invokable method
+     * @throws IllegalStateException if this method is called more than once with the same {@code position}
+     */
+    InvokerBuilder<T> setArgumentTransformer(int position, Class<?> clazz, String methodName);
+
+    /**
+     * Configures an output transformer for the return value.
+     *
+     * @param clazz class that declares the transformer
+     * @param methodName transformer method name
+     * @return this builder
+     * @throws IllegalStateException if this method is called more than once
+     */
+    InvokerBuilder<T> setReturnValueTransformer(Class<?> clazz, String methodName);
```

Comment:
These seem too stringly typed and dynamic, can we not use method handles and provide generics for the return type? Why add a requirement for reflection here?

Reply:
It is rather stringly typed, but all alternatives I could think of are as much stringly typed if not more. The idea here is to type check what we can during deployment (which would be during application build for build-time implementations) and let the JVM to sort out the rest. To allow transformers to convert values from one type to another, which seems like a hard requirement to me, we can only type check the output type for input transformers (as the output of an input transformer is passed to the invokable method and hence its type must match) and the input type for output transformers (because the input of an output transformer is obtained from the invokable method and hence its type must match).

Or did you mean the syntax of referring to the transformer method, the tuple `(Class, String)`? I considered 2 alternatives:

1. Require transformers to implement some interface. Not nice, because it requires extra allocations and virtual dispatch. I wanted transformers to be `static` methods mainly to eliminate overhead, but it's also nice in that it tells users that transformers should be stateless. (They can also be instance methods, but that is [or should be] by construction restricted to simple scenarios where the `static` method would just call that instance method.)
2. Use method references (as in, `SomeClass::transformer`). This would require defining functional interfaces for all possible signatures of transformers and provide overloads that accept these functional interfaces. It would also require some way to find the method during deployment (that is, during application build for build-time implementations). Apparently there's a trick to do that if the functional interface extends `Serializable`, which seems strange to me. Finally, we might also want to detect (and forbid?) situation where the caller doesn't pass us a method reference, but an actual lambda.

I'd really love to use method references, but I couldn't figure out how to implement it in a reasonable way. So I settled with just a pair `(Class, String)`, which is basically what I'd need from the method reference, except explicit (and unfortunately not checked during compilation).

Finally, the idea is that all invokers are built during deployment -- there's no way to obtain an `InvokerBuilder` during application runtime. Therefore, there's no requirement for reflection here -- build-time implementations are supposed to find the transformer method using whatever language model they use (e.g. Jandex in case of Quarkus) and build the invoker using whatever mechanism they prefer (e.g. bytecode generation in case of Quarkus).

Reply:
ok I am also wondering though why all of this transformation and decoration is necessary? Surely this change to the spec should be purely focused on executable methods and the transformation / decoration side should be in the interceptors spec?

Reply:
> the transformation / decoration side should be in the interceptors spec?

That was also my initial question when I saw Ladislav's prototype.
However, given the use cases above, I tend to think this is better approach. The idea is that invokeable methods are a feature used by extensions (or integrators, if you will) and as such the integrator might want to do their own specific tweak in how the method is invoked and which parameters are looked up and so on. This would be hard to achieve via interception as multiple extensions trying to handle it differently might easily clash.
Furthermore, unlike interceptors, in this case there is always at most one such transformation per registered invoker (or per extension/integrator that wants to tweak it specifically) - it also doesn't need explicit enablement and doesn't deal with priority meaning it isn't subject to interference from ordering. Last but not least, you can still apply other interceptors onto the method independently of what each extension decides to perform with the invocation.

Reply:
> Frameworks often want to let users use multiple signatures of the methods the framework will invoke, but also want to handle them uniformly internally.

Would it be useful then to allow the user to supply a functional interface which defines the uniform shape they're trying to create, and allow the `InvokerBuilder` to return an instance of that interface instead of an `Invoker`? 

Would that allow for better type checking?

> the syntax of referring to the transformer method, the tuple (Class, String)

I must stress that as a Java developer using regular Java without any ahead of time compilation, this seems absolutely horrible. Any transformer seems like it should be a `Function<I, O>` and accept a lambda.

I assume that this is a limitation of doing stuff at build time and I'd need to know more about what limitations that imposes to be able to suggest something better, but I'd hope that we can create something which doesn't seem fully alien to a regular Java developer.

Lastly, on a separate note, I would want the Javadoc for the methods which set a transformer to include the allowed signatures of the transformer method. It's currently listed in the class javadoc, but that information should be referenced or repeated in the method doc.

Reply:
Some good points there. Indeed I chose the `(Class, String)` representation of method references to enable generating built invokers during application build. I'm not terribly fond of it, but it's reasonably simple and I'm going to claim that anyone writing CDI extensions is very likely to have at least _some_ experience with the Java reflection API, so this will hardly be fully alien to them. Using lambdas probably makes generating invokers during build harder, if not impossible, but I didn't really try it, so take that with a grain of salt.

When it comes to Javadoc, I expect that a huge part of what's there currently will move to the specification text and the javadoc will only contain the more practical parts (and examples). That includes the possible signatures of transformers, if we stay with `(Class, String)`, listed directly on the builder methods, but for now, I wanted to keep that on a single place.

Reply:
what I don't like here is the assumption that reflection will be used. The `Class` parameter should be `ClassInfo`

Reply:
This is a user facing API - how is user going to get `ClassInfo` object for their arbitrary `Foo` class? That class doesn't even need to be a bean, so you cannot get that from another BCE event either.
And as far as I can tell, we don't have a way to create that in the lang model API either.

If we wanted to go without `Class`, we'd probably need to go full `String` mode which is pretty ugly...

Reply:
So the `Class` parameter does not mean reflection needs to be used. It is a possible implementation, but not the only one. The API is designed so that it can be implemented purely at build time. I actually have a fairly complete implementation of this API in ArC, where all invokers created through the `InvokerBuilder` are generated during build. In such implementation, the `Class` object is only used to obtain the binary name of the class, nothing else. I'd be happy to present how my implementation works, and I'd of course love to hear if this API contains any obstacles for your implementation.

I'll experiment with using method references, but I fear that it isn't exactly easy to extract the necessary information from them during build time, as a method reference is indistinguishable from a lambda expression.

Reply:
Actually thinking about it, I see one possible problem for your implementation (based on what I know -- if you could confirm, that would be great). The specification currently assumes that an implementation can inspect the class so that it can:

- find the named method, and use its signature for code generation,
- do the limited type checking the specification prescribes.

I didn't notice any issue with that, because our implementation has full access to the deployment. That isn't exactly the case for your implementation, if I remember correctly.

Now, indeed one way to solve it would be to use reflection to inspect the class. That would be reflection during build time, not at runtime -- but this might still be problematic. I'm no expert here, so please correct me if I'm wrong.

I assume if we could pull off some trick with method references, that would fix the issue for you, as the method reference would come with the full signature and you wouldn't have to search for the method on the class. Is that right?

Reply:
so I see there is `InvokerBuilder<InvokerInfo> createInvoker(MethodInfo method);` so it is probably ok.